pkg/retry/retry.go

```diff
@@ -77,3 +85,80 @@ func OnError(
 		}
 	}
 }
+
+// Call executes a function repeatedly until it succeeds or the retry strategy
+// indicates that no more retries should be attempted.
+//
+// If no retry strategy is provided, it defaults to an exponential backoff strategy
+// with 5 max retries, 1000ms initial delay, and 16000ms max delay.
+//
+// Returns the result from the work function and any error that occurred if retries are exhausted.
+func Call[T any](
+	work func() (T, error),
+	retryStrategy ...RetryStrategyFunc,
+) (T, error) {
+	var (
+		result T
+		err    error
+	)
+	if retryStrategy == nil {
+		retryStrategy = []RetryStrategyFunc{DefaultExponentialDelay}
+	}
+
+	for retryCount := 0; ; retryCount++ {
+		result, err = work()
+
+		// Stop retrying and return the result if no error occurred
+		if err == nil {
+			return result, err
+		}
+
+		// Stop retrying and return the result if the error is non-retryable
+		if ErrNonRetryable.Is(err) {
+			return result, err
+		}
+
+		// Stop retrying and return if the error is a gRPC error
```

Comment:
:ok_hand: Nice :sunglasses: 

I think we should make an exception for ["unavailable"](https://pkg.go.dev/google.golang.org/grpc/codes@v1.67.1#Unavailable) and perhaps ["resource exhausted"](https://pkg.go.dev/google.golang.org/grpc/codes@v1.67.1#ResourceExhausted) codes though because they both represent potentially transient errors.